### PR TITLE
Fix JCasC backwards compatibility

### PR DESCRIPTION
**Pull request by res0nance**
- Fix backwards compatibility with JCasC configurations broken by #329 
- Add tests for JCasC

### CODE REVIEW

File: src/main/java/hudson/plugins/ec2/SlaveTemplate.java [hunk 1]
```diff
@@ -206,15 +206,19 @@ public SlaveTemplate(String ami, String zone, SpotConfiguration spotConfig, Stri
         this.tags = tags;
         this.idleTerminationMinutes = idleTerminationMinutes;
         this.associatePublicIp = associatePublicIp;
-        this.connectionStrategy = connectionStrategy;
+        if (connectionStrategy == null) {
```
Comment (jvz) [inline src/main/java/hudson/plugins/ec2/SlaveTemplate.java]: You could change all this line to just `this.connectionStrategy = connectionStrategy == null ? ConnectionStrategy.PRIVATE_IP : connectionStrategy;`

File: src/test/java/hudson/plugins/ec2/ConfigurationAsCodeTest.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Rule
+    public JenkinsConfiguredWithCodeRule j = new JenkinsConfiguredWithCodeRule();
+
+    @Test
+    @ConfiguredWithCode("EC2CloudEmpty.yml")
+    public void testEmptyConfig() throws Exception {
+        final AmazonEC2Cloud ec2Cloud = (AmazonEC2Cloud) Jenkins.get().getCloud("ec2-");
+        assertNotNull(ec2Cloud);
+        assertEquals(ec2Cloud.getTemplates().size(), 0);
```
Comment (jvz) [inline src/test/java/hudson/plugins/ec2/ConfigurationAsCodeTest.java]:
```suggestion
        assertEquals(0, ec2Cloud.getTemplates().size());
```

These assertions are usually ordered `assertEquals(expected, actual)`.

**Comment by res0nance** [comment]
@casz are there any other cases of JCasC we should test for?

**Comment by casz** [comment]
I don't actively use ec2 😓 But let me ask our gitter users 👍

**Comment by casz** [comment]
Here is one exception log and yaml file from user:
https://gist.github.com/Shaps/efb4aed61955cd8301e2ae268d90dee6
https://gist.github.com/Shaps/be5503ec902d4ce0dbe4958d163b6902

**Comment by timja** [comment]
Model export test docs:
https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/PLUGINS.md#model-export-test

**Comment by res0nance** [comment]
Thanks for your input guys I'll add an export test in a separate PR. I'll look into that yml failure as well

**Comment by devblueray** [comment]
https://gist.github.com/devblueray/6980c4dfd3235f9ed5c2a71d1f46c4d5

Here's a gist of the error I am getting with the EC2 Plugin

**Comment by res0nance** [comment]
@casz I got the yaml you linked working and added a MRE test case for it. Wondering if this was the right way to solve it.

@devblueray Thanks for the report do you have a `yml` file that will reproduce this issue.

Overall this PR is getting too big. I'll implement all the export stuff in a separate PR

**Comment by devblueray** [comment]
I don't, this was from the export.  I am attempting to write the relevant config sections by hand to see if they work.

**Comment by res0nance** [comment]
Alright I'll add an export test in a separate PR to deal with export issues. This PR is primarily to deal with the import issues

**Comment by oleg-nenashev** [comment]
@res0nance Hi. What would be the release ETA?

**Comment by res0nance** [comment]
Hey @oleg-nenashev I'm not a maintainer so I cant release it but #349, #296 and #360 probably should be merged before release. @thoulen does the releases so he would be able to speak about that.

**Pull request merged**